ConcurrentModificationException
===============================================

This exception may be thrown by methods that have detected concurrent modification of an object when such modification is not permissible. [Javadoc  

https://docs.oracle.com/javase/7/docs/api/java/util/ConcurrentModificationException.html]

Note that this exception does not always indicate that an object has been concurrently modified by a different thread. If a single thread issues a sequence of method invocations that violates the contract of an object, the object may throw this exception. For example, if a thread modifies a collection directly while it is iterating over the collection with a fail-fast iterator, the iterator will throw this exception.






Why iterator.remove does not throw ConcurrentModificationException
https://stackoverflow.com/questions/24856811/why-iterator-remove-does-not-throw-concurrentmodificationexception#:~:text=ConcurrentModificationException%20is%20not%20thrown%20by,modify%20an%20collection%20while%20iterating.&text=Removes%20from%20the%20underlying%20collection,per%20call%20to%20next().